src/shims/unix/android/thread.rs

```diff
@@ -0,0 +1,72 @@
+use std::iter;
+
+use rustc_span::Symbol;
+use rustc_target::spec::abi::Abi;
+
+use crate::helpers::check_min_arg_count;
+use crate::shims::unix::thread::{DEFAULT_THREAD_NAME, EvalContextExt as _};
+use crate::*;
+
+pub fn prctl<'tcx>(
+    this: &mut MiriInterpCx<'tcx>,
+    link_name: Symbol,
+    abi: Abi,
+    args: &[OpTy<'tcx>],
+    dest: &MPlaceTy<'tcx>,
+) -> InterpResult<'tcx> {
+    // We do not use `check_shim` here because `prctl` is variadic. The argument
+    // count is checked bellow.
+    this.check_abi_and_shim_symbol_clash(abi, Abi::C { unwind: false }, link_name)?;
+
+    // FIXME: Use PR_SET_NAME and PR_GET_NAME constants when
+    // https://github.com/rust-lang/libc/pull/3941 lands.
+    const PR_SET_NAME: i32 = 15;
+    const PR_GET_NAME: i32 = 16;
+    const TASK_COMM_LEN: usize = 16;
+
+    let [op] = check_min_arg_count("prctl", args)?;
+    let res = match this.read_scalar(op)?.to_i32()? {
+        PR_SET_NAME => {
+            let [_, name] = check_min_arg_count("prctl(PR_SET_NAME, ...)", args)?;
+
+            let name = this.read_scalar(name)?.to_pointer(this)?;
+            let name =
+                this.read_c_str(name)?.iter().take(TASK_COMM_LEN - 1).copied().collect::<Vec<_>>();
+
+            let thread = this.pthread_self()?.to_int(this.libc_ty_layout("pthread_t").size)?;
+            let thread = ThreadId::try_from(thread).unwrap();
+
+            this.set_thread_name(thread, name);
+            Scalar::from_u32(0)
+        }
+        PR_GET_NAME => {
+            let [_, name] = check_min_arg_count("prctl(PR_GET_NAME, ...)", args)?;
+
+            let name_out = this.read_scalar(name)?;
+            let name_out = name_out.to_pointer(this)?;
+
+            let thread = this.pthread_self()?.to_int(this.libc_ty_layout("pthread_t").size)?;
+            let thread = ThreadId::try_from(thread).unwrap();
+
+            // FIXME: we should use the program name if the thread name is not set
+            let name = this.get_thread_name(thread).unwrap_or(DEFAULT_THREAD_NAME).to_owned();
+            let name_len = name.len().max(TASK_COMM_LEN - 1);
+
+            this.eval_context_mut().write_bytes_ptr(
```

Comment:
The reason why `write_c_str` isn't used here is to follow up the `prctl` behavior, and let `write_bytes_ptr` to verify the destination length to see if the buffer doesn't match `TASK_COMM_LEN`.

Reply:
`write_c_str` can do that, though?

Reply:
You should be able to do something like `write_c_str` on `&name[..name_len]`.

Reply:
Yes, it can do that, but it has a `size` check inside it which has no meaning in the current situation (the `name` is truncated already). Anyway, I will follow up your suggestion.

Reply:
Still better than duplicating the null termination handling.

Reply:
Actually, it also requires changes to the input string because `write_c_str` requires the source and destination buffers sizes to have the same length. Then it should collected into a `Vec` with appended `0u8` to it, and only then to be passed `write_c_str` which internally will do the check, append one more `0u8` and only then call `write_bytes_ptr`.

Reply:
Rewrote it using `write_c_str`. There's no null terminator handling, but additional nulls appended at the end.

Reply:
Why are you appending additional nulls at the end?

Reply:
That's due to these lines ([source](https://github.com/rust-lang/rust/blob/baaf3e65abab5a281715230bcb1785610629ae7d/compiler/rustc_const_eval/src/interpret/memory.rs#L1281-L1286)):

```rust
// `zip` would stop when the first iterator ends; we want to definitely
// cover all of `bytes`.
for dest in bytes {
    *dest = src.next().expect("iterator was shorter than it said it would be");
}
assert_matches!(src.next(), None, "iterator was longer than it said it would be");
```

The destination and source buffers must have exactly the same size, and `PR_GET_NAME` requires the buffer to be exactly 16 bytes long ([docs](https://www.man7.org/linux/man-pages/man2/PR_GET_NAME.2const.html)):

```c
int prctl(PR_GET_NAME, const char name[16]);
```

In contrast, `pthread_getname_np` accepts the buffer length ([source](https://www.man7.org/linux/man-pages/man3/pthread_getname_np.3.html)):

```c
int pthread_getname_np(pthread_t thread, char name[.size], size_t size);
```

Though, it doesn't mean that it will return a name back because the `size` must be at least like `TASK_COMM_LEN` or even longer.

```
The pthread_getname_np() function can fail with the following
error:

ERANGE The buffer specified by name and size is too small to hold
       the thread name.
```

The code from `bionic` (I don't have another `libc` implementation openned right now, so let's use it as an example) demonstrates that check ([source](https://android.googlesource.com/platform/bionic/+/8a18af52d9b9344497758ed04907a314a083b204/libc/bionic/pthread_setname_np.cpp)):

```c
__BIONIC_WEAK_FOR_NATIVE_BRIDGE
int pthread_getname_np(pthread_t t, char* buf, size_t buf_size) {
  ErrnoRestorer errno_restorer;
  if (buf_size < MAX_TASK_COMM_LEN) return ERANGE;
```

I also assume that you would ask why `pthread_getname_np` cannot be used in `std` instead of `prctl`, so the question is it requires `bionic` being compiled with `_GNU_SOURCE` defined. Presence of `pthread_getname_np` in `bionic` without `_GNU_SOURCE` is just a bug. So, no it cannot be use (:

I feel that I'm close to mind frapping if it hasn't started yet, so fixing the tests is a task for tomorrow.

Reply:
Forgot to mention, according to the explanation above, you have `pthread_getname_np` implemented wrongly. It must accept only `TASK_COMM_LEN` sized buffers and reject others with `ERANGE`. I will fix that too.

Reply:
I would say for a thread name of length N < 15, `prctl` only writes the first N+1 bytes and leaves the rest untouched. There is *a* null terminator, but why do you want to fill *the entire rest of the buffer* with NULL? That's not usually how we implement APIs like this.

Reply:
> There is a null terminator, but why do you want to fill the entire rest of the buffer with NULL? That's not usually how we implement APIs like this.

It was my tired mind doing shit. I read the `miri`'s code incorrectly, so my picture was that the `write_bytes_ptr` requires filling the whole allocated buffer. That's correct, but an earlier call to `get_ptr_alloc_mut` request the exact amount of memory to be borrowed, and it throws if it goes out of the bounds.